test/multiverse/suites/roda/roda_instrumentation_test.rb

```diff
@@ -0,0 +1,146 @@
+    RodaTestApp
+  end
+
+  def test_nil_verb
+    NewRelic::Agent::Instrumentation::Roda::TransactionNamer.stub(:http_verb, nil) do
+      get('/home')
+      txn = harvest_transaction_events![1][0]
+
+      assert_equal 'Controller/Roda/RodaTestApp/home', txn[0]['name']
+      assert_equal 200, txn[2][:'http.statusCode']
+    end
+  end
+
+  def test_http_verb_request_no_request_method
+    fake_request = Struct.new('FakeRequest', :path).new
+    name = NewRelic::Agent::Instrumentation::Roda::TransactionNamer.transaction_name(fake_request)
+
+    assert_equal ::NewRelic::Agent::UNKNOWN_METRIC, name
+  end
+
+  def test_request_is_recorded
+    get('/home')
+    txn = harvest_transaction_events![1][0]
+
+    assert_equal 'Controller/Roda/RodaTestApp/GET home', txn[0]['name']
+    assert_equal 200, txn[2][:'http.statusCode']
+  end
+
+  def test_500_response_status
+    get('/error')
+    errors = harvest_error_traces!
+    txn = harvest_transaction_events!
+
+    assert_equal 500, txn[1][0][2][:"http.statusCode"]
+    assert_equal 'Oh No!', last_response.body
+    assert_equal 1, errors.size
+  end
+
+  def test_404_response_status
+    get('/nothing')
+    errors = harvest_error_traces!
+    txn = harvest_transaction_events!
+
+    assert_equal 404, txn[1][0][2][:"http.statusCode"]
+    assert_equal 0, errors.size
+  end
+
+  def test_empty_route_name_and_response_status
+    get('')
+    errors = harvest_error_traces!
+    txn = harvest_transaction_events![1][0]
+
+    assert_equal 'Controller/Roda/RodaTestApp/GET /', txn[0]['name']
+    assert_equal 302, txn[2][:'http.statusCode']
+  end
+
+  def test_roda_auto_middleware_disabled
+    with_config(:disable_roda_auto_middleware => true) do
+      RodaNoMiddleware.build_rack_app_with_tracing {}
+
+      assert_truthy NewRelic::Agent::Agent::config[:disable_roda_auto_middleware]
+    end
+  end
+
+  def test_roda_instrumentation_works_if_middleware_disabled
+    with_config(:disable_middleware_instrumentation => true) do
+      get('/home')
+      txn = harvest_transaction_events![1][0]
+
+      assert_equal 'Controller/Roda/RodaTestApp/GET home', txn[0]['name']
+    end
+  end
+
+  def test_transaction_name_error
+    NewRelic::Agent.stub(:logger, NewRelic::Agent::MemoryLogger.new) do
+      # pass in {} to produce an error, because {} doesn't support #path and
+      # confirm that the desired error handling took place
+      result = NewRelic::Agent::Instrumentation::Roda::TransactionNamer.transaction_name({})
+
+      assert_equal NewRelic::Agent::UNKNOWN_METRIC, result
+      assert_logged(/NoMethodError.*Error encountered trying to identify Roda transaction name/)
+    end
+  end
+
+  def test_rack_request_params_error
+    NewRelic::Agent.stub(:logger, NewRelic::Agent::MemoryLogger.new) do
+      # Have the #params call made on the request raise an exception to test
+      # the error handling
+      RodaTestApp::RodaRequest.any_instance.stubs(:params).raises(StandardError.new)
```

Comment:
Following up on offline discussions on this one...

- Minitest doesn't support (and one might even infer it discourages) the stubbing of any instance
- The `any_instance` here is provided by the Mocha testing gem
- Personally I think introducing Mocha to a Minitest based project detracts from the Minitest experience as much as it adds to it. Perhaps it's like adding a sidecar to a motorcycle. It's clear that the sidecar offers a distinct advantage over a motorcycle without a sidecar. But there are also fundamental changes to the motorcycling experience that could be considered as drawbacks.
- Testing any instance is well known anti-pattern given the risk of having stubbing behavior bleed over to other instances in play in a testing context where the stubbing behavior would be problematic.
    - The risk is mitigated here in two ways:
        - We are assured of only having one Rack request in play during the test.
        - We do not (currently) make use of parallelized, concurrent tests.
- Typically the best practice is to have a handle on the specific instance under test. The need to reach for the stubbing of any instance could itself indicate complexity of code. This complexity might come from not following TDD or could indicate a focus / prioritization on performance over testability. For whatever reason, tests that require stubbing any instance might indicate difficulties in maintainability.
- We use this `any_instance` method quite a lot in this project. It is completely acceptable to use it again here from a PR approval standpoint. The maintainers may at one point decide to dissuade future usage or refactor out existing usage of `any_instance`, but for now there's a green light.
    - We bring in a RuboCop gem specific to Minitest, but unfortunately given that `any_instance` comes from Mocha, its usage is not flagged by any cops.
- Looking at the context for this exact line of code, the benefit of using `any_instance` is pretty clear. We invoke the `Rack::Test` `get` method, which via `Forwardable` goes through the `Rack::Test` flow of calling the user defined `app` method to get an app class. An instance of that class is instantiated and the request params object is prepped. For us to be able to get at the exact single instance of the class (as opposed to any instance) that will field the web request, we would need to change the overall flow of this test and have it be different from all the others. Perhaps we could stub `new` on the Rack app and use that stubbing for a foothold while retaining the Rack test integration style of using `get`. Alternatively, we could have this test not leverage `Rack::Test` and we could instead directly invoke the `rack_request_params` method being tested. (We'd use a unit test instead of an integration test). The benefits of having this test look and act like the others is a pretty clear advantage of `any_instance`.

Reply:
Thanks for the detailed thoughts—loving the motorcycle analogy. I agree it's weird to mix Mocha and Minitest. I've updated this and to go with the unit-style test— just calling `rack_request_params` directly, which is going to fail since no Rack request will be in existence. 

Updating our codebase to get rid of `ay_instance` would be a fun project. I think it would help with learning.